ports/nrf/drivers/usb/usb_cdc.c

```diff
@@ -0,0 +1,228 @@
+uint8_t tx_ringbuf_array[4096];
+uint8_t rx_buf[1];
+uint8_t tx_buf[1];
+volatile ringbuf_t rx_ringbuf;
+volatile ringbuf_t tx_ringbuf;
+
+static void board_init(void) {
+    // Config clock source.
+#ifndef BLUETOOTH_SD
+    NRF_CLOCK->LFCLKSRC = (uint32_t)((CLOCK_LFCLKSRC_SRC_Xtal << CLOCK_LFCLKSRC_SRC_Pos) & CLOCK_LFCLKSRC_SRC_Msk);
+    NRF_CLOCK->TASKS_LFCLKSTART = 1UL;
+#endif
+
+    // Priorities 0, 1, 4 (nRF52) are reserved for SoftDevice
+    // 2 is highest for application
+    NRFX_IRQ_PRIORITY_SET(USBD_IRQn, 2);
+
+    // USB power may already be ready at this time -> no event generated
+    // We need to invoke the handler based on the status initially
+    uint32_t usb_reg;
+
+#ifdef BLUETOOTH_SD
+    uint8_t sd_en = false;
+    sd_softdevice_is_enabled(&sd_en);
+
+    if (sd_en) {
+        sd_power_usbdetected_enable(true);
+        sd_power_usbpwrrdy_enable(true);
+        sd_power_usbremoved_enable(true);
+
+        sd_power_usbregstatus_get(&usb_reg);
+    } else
+#endif
+    {
+        // Power module init
+        const nrfx_power_config_t pwr_cfg = { 0 };
+        nrfx_power_init(&pwr_cfg);
+
+        // Register tusb function as USB power handler
+        const nrfx_power_usbevt_config_t config = { .handler = (nrfx_power_usb_event_handler_t) tusb_hal_nrf_power_event };
+        nrfx_power_usbevt_init(&config);
+
+        nrfx_power_usbevt_enable();
+
+        usb_reg = NRF_POWER->USBREGSTATUS;
+    }
+
+    if (usb_reg & POWER_USBREGSTATUS_VBUSDETECT_Msk) {
+        tusb_hal_nrf_power_event(NRFX_POWER_USB_EVT_DETECTED);
+    }
+
+#ifndef BLUETOOTH_SD
+    if (usb_reg & POWER_USBREGSTATUS_OUTPUTRDY_Msk) {
+        tusb_hal_nrf_power_event(NRFX_POWER_USB_EVT_READY);
+    }
+#endif
+}
+
+static bool cdc_rx_any(void) {
+    return rx_ringbuf.iput != rx_ringbuf.iget;
+}
+
+static int cdc_rx_char(void) {
+    return ringbuf_get((ringbuf_t*)&rx_ringbuf);
+}
+
+static bool cdc_tx_any(void) {
+    return tx_ringbuf.iput != tx_ringbuf.iget;
+}
+
+static int cdc_tx_char(void) {
+    return ringbuf_get((ringbuf_t*)&tx_ringbuf);
+}
+
+static void cdc_task(void)
+{
+    if ( tud_cdc_connected() ) {
+        // connected and there are data available
+        while (tud_cdc_available()) {
+            int c;
+            uint32_t count = tud_cdc_read(&c, 1);
+            (void)count;
+            ringbuf_put((ringbuf_t*)&rx_ringbuf, c);
+        }
+
+        int chars = 0;
+        while (cdc_tx_any()) {
+            if (chars < 64) {
+               tud_cdc_write_char(cdc_tx_char());
+               chars++;
+            } else {
+               chars = 0;
+               tud_cdc_write_flush();
+            }
+        }
+
+        tud_cdc_write_flush();
+    }
+}
+
+static void usb_cdc_loop(void) {
+    tud_task();
+    cdc_task();
+}
+
+int usb_cdc_init(void)
+{
+    static bool initialized = false;
+    if (!initialized) {
+
+#if BLUETOOTH_SD
+        // Initialize the clock and BLE stack.
+        ble_drv_stack_enable();
```

Comment:
The mentioned hack to get USB events propagated in case of compilation with SD=s140.